File: regions/system/library/go/building-blocks/inmemory_statestore.go
package buildingblocks

import (
	"context"
	"fmt"
	"strconv"
	"sync"
	"sync/atomic"
)

type stateEntry struct {
	value []byte
	etag  string
}

// Compile-time interface compliance check.
var _ StateStore = (*InMemoryStateStore)(nil)

// InMemoryStateStore is an in-memory implementation of StateStore for testing.
type InMemoryStateStore struct {
	mu      sync.RWMutex
	entries map[string]*stateEntry
	counter atomic.Uint64
	status  ComponentStatus
}

// NewInMemoryStateStore creates a new InMemoryStateStore.
func NewInMemoryStateStore() *InMemoryStateStore {
	return &InMemoryStateStore{
		entries: make(map[string]*stateEntry),
		status:  StatusUninitialized,
	}
}

func (s *InMemoryStateStore) Name() string    { return "inmemory-statestore" }
func (s *InMemoryStateStore) Version() string { return "1.0.0" }

func (s *InMemoryStateStore) Init(_ context.Context, _ Metadata) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.status = StatusReady
	return nil
}

func (s *InMemoryStateStore) Close(_ context.Context) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.status = StatusClosed
	return nil
}

func (s *InMemoryStateStore) Status(_ context.Context) ComponentStatus {
	s.mu.RLock()
	defer s.mu.RUnlock()
	return s.status
}

func (s *InMemoryStateStore) nextETag() string {
	return strconv.FormatUint(s.counter.Add(1), 10)
}

// Get returns the state entry for key, or nil if not found.
func (s *InMemoryStateStore) Get(_ context.Context, key string) (*StateEntry, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	e, ok := s.entries[key]
	if !ok {
		return nil, nil
	}
	return &StateEntry{Key: key, Value: e.value, ETag: &ETag{Value: e.etag}}, nil
}

// Set stores a value. If req.ETag is non-nil, it enforces optimistic concurrency.
func (s *InMemoryStateStore) Set(_ context.Context, req *SetRequest) (*ETag, error) {
	s.mu.Lock()
	defer s.mu.Unlock()
	existing, ok := s.entries[req.Key]
	if req.ETag != nil {
		if !ok {
			return nil, &ETagMismatchError{Key: req.Key, Expected: req.ETag, Actual: nil}
		}
		if existing.etag != req.ETag.Value {
			return nil, &ETagMismatchError{Key: req.Key, Expected: req.ETag, Actual: &ETag{Value: existing.etag}}
		}
	}
	newETag := s.nextETag()
	s.entries[req.Key] = &stateEntry{value: req.Value, etag: newETag}
	return &ETag{Value: newETag}, nil
}

// Delete removes the entry for key. If etag is non-nil, it enforces optimistic concurrency.
func (s *InMemoryStateStore) Delete(_ context.Context, key string, etag *ETag) error {
	s.mu.Lock()
	defer s.mu.Unlock()
	existing, ok := s.entries[key]
	if !ok {
		return nil
	}
	if etag != nil && existing.etag != etag.Value {
		return &ETagMismatchError{Key: key, Expected: etag, Actual: &ETag{Value: existing.etag}}
	}
	delete(s.entries, key)
	return nil
}

// BulkGet retrieves multiple entries by key.
func (s *InMemoryStateStore) BulkGet(ctx context.Context, keys []string) ([]*StateEntry, error) {
	results := make([]*StateEntry, 0, len(keys))
	for _, key := range keys {
		entry, err := s.Get(ctx, key)
		if err != nil {
			return nil, fmt.Errorf("bulk get key %q: %w", key, err)
		}
		results = append(results, entry)
	}
	return results, nil
}

// BulkSet stores multiple entries in sequence.
func (s *InMemoryStateStore) BulkSet(ctx context.Context, requests []*SetRequest) ([]*ETag, error) {
	etags := make([]*ETag, 0, len(requests))
	for _, req := range requests {
		etag, err := s.Set(ctx, req)
		if err != nil {
			return nil, fmt.Errorf("bulk set key %q: %w", req.Key, err)
		}
		etags = append(etags, etag)
	}
	return etags, nil
}

File: regions/system/library/typescript/building-blocks/src/inmemory_statestore.ts
// インメモリステートストア実装モジュール。
// テストおよびローカル開発環境向けに、Redis等の外部ストア不要でStateStoreを提供する。

import { ETagMismatchError } from './errors.js';
import type { ComponentStatus } from './component.js';
import type { StateEntry, StateStore } from './statestore.js';

/**
 * 内部ストアエントリーの構造を定義するインターフェース。
 * バイナリ値とETagをセットで管理することで楽観的並行制御を実現する。
 */
interface _StoreEntry {
  // 保存されたバイナリ値
  value: Uint8Array;
  // 楽観的並行制御に使用するETag文字列
  etag: string;
}

// ETagの一意性を保証するためのモジュールスコープカウンター
let _etagCounter = 0;

/**
 * インメモリステートストア実装クラス。
 * キーとバイナリ値のMapで状態を管理し、ETagによる楽観的並行制御をサポートする。
 * 外部ストアへの接続なしにステート管理ロジックをテストできる。
 */
export class InMemoryStateStore implements StateStore {
  // コンポーネントの識別名
  readonly name: string;
  // コンポーネント種別（ステートストアであることを示す）
  readonly componentType = 'statestore';

  // 現在のコンポーネント状態（未初期化・準備完了・クローズ済み）
  private _status: ComponentStatus = 'uninitialized';
  // キー → エントリー（値+ETag）のマッピングを保持するMap
  private readonly _entries = new Map<string, _StoreEntry>();

  // デフォルト名を指定してインスタンスを生成する
  constructor(name = 'inmemory-statestore') {
    this.name = name;
  }

  // コンポーネントを初期化し、状態をreadyに変更する
  async init(): Promise<void> { this._status = 'ready'; }

  // コンポーネントをクローズし、全エントリーをクリアして状態をclosedに変更する
  async close(): Promise<void> { this._entries.clear(); this._status = 'closed'; }

  // 現在のコンポーネント状態を返す
  async status(): Promise<ComponentStatus> { return this._status; }

  // バックエンドを示すメタデータを返す
  metadata(): Record<string, string> { return { backend: 'memory' }; }

  /**
   * 指定キーのステートエントリーを取得して返す。
   * キーが存在しない場合はnullを返す。
   */
  async get(key: string): Promise<StateEntry | null> {
    const entry = this._entries.get(key);
    // エントリーが存在しない場合はnullを返す
    if (!entry) return null;
    // キー・値・ETagを含むStateEntryオブジェクトを返す
    return { key, value: entry.value, etag: entry.etag };
  }

  /**
   * 指定キーに値を保存し、新しいETagを返す。
   * etagが指定された場合は楽観的並行制御を行い、不一致の場合はETagMismatchErrorをスローする。
   * etagが未指定の場合は無条件で上書きする。
   */
  async set(key: string, value: Uint8Array, etag?: string): Promise<string> {
    const existing = this._entries.get(key);
    // ETagが指定されている場合は楽観的並行制御チェックを実行する
    if (etag !== undefined) {
      // エントリーが存在しない場合はETag不一致エラーをスローする
      if (!existing) throw new ETagMismatchError(key, etag, '');
      // 現在のETagと異なる場合は競合エラーをスローする
      if (existing.etag !== etag) throw new ETagMismatchError(key, etag, existing.etag);
    }
    // カウンターをインクリメントして新しいETagを生成する
    const newETag = String(++_etagCounter);
    this._entries.set(key, { value, etag: newETag });
    return newETag;
  }

  /**
   * 指定キーのエントリーを削除する。
   * キーが存在しない場合は何もしない。
   * etagが指定された場合は楽観的並行制御を行い、不一致の場合はETagMismatchErrorをスローする。
   */
  async delete(key: string, etag?: string): Promise<void> {
    const existing = this._entries.get(key);
    // エントリーが存在しない場合は何もしない
    if (!existing) return;
    // ETagが指定されており、現在のETagと異なる場合は競合エラーをスローする
    if (etag !== undefined && existing.etag !== etag) {
      throw new ETagMismatchError(key, etag, existing.etag);
    }
    this._entries.delete(key);
  }

  /**
   * 複数キーのステートエントリーをまとめて取得して返す。
   * 存在するエントリーのみを結果に含め、存在しないキーはスキップする。
   */
  async bulkGet(keys: string[]): Promise<StateEntry[]> {
    const results: StateEntry[] = [];
    // 各キーのエントリーを順次取得し、存在するものだけをresultsに追加する
    for (const key of keys) {
      const entry = await this.get(key);
      if (entry) results.push(entry);
    }
    return results;
  }

  /**
   * 複数のキーと値のペアをまとめて保存し、それぞれの新しいETagを返す。
   * ETagを指定しない無条件上書きで保存するため、競合チェックは行わない。
   */
  async bulkSet(entries: Array<{ key: string; value: Uint8Array }>): Promise<string[]> {
    const etags: string[] = [];
    // 各エントリーを順次保存し、生成されたETagをetags配列に追加する
    for (const { key, value } of entries) {
      etags.push(await this.set(key, value));
    }
    return etags;
  }
}
